bigtable/bigtable.go

```diff
@@ -331,70 +346,248 @@ func (r RowList) retainRowsAfter(lastRowKey string) RowSet {
+	case rangeUnbounded:
+		break
+	}
+
+	switch r.endBound {
+	case rangeOpen:
+		contains = contains && r.end > row
+		break
+	case rangeClosed:
+		contains = contains && r.end >= row
+		break
+	case rangeUnbounded:
+		break
+	}
+
+	return contains
 }
 
 // String provides a printable description of a RowRange.
 func (r RowRange) String() string {
-	a := strconv.Quote(r.start)
-	if r.Unbounded() {
-		return fmt.Sprintf("[%s,∞)", a)
+	var startStr string
+	switch r.startBound {
+	case rangeOpen:
+		startStr = "(" + strconv.Quote(r.start)
+		break
+	case rangeClosed:
+		startStr = "[" + strconv.Quote(r.start)
+		break
+	case rangeUnbounded:
+		startStr = "(∞"
+		break
 	}
-	return fmt.Sprintf("[%s,%q)", a, r.limit)
+
+	var endStr string
+	switch r.endBound {
+	case rangeOpen:
+		endStr = r.end + ")"
+		break
+	case rangeClosed:
+		endStr = r.end + "]"
+		break
+	case rangeUnbounded:
+		endStr = "∞)"
+		break
+	}
+
+	return fmt.Sprintf("%s,%s", startStr, endStr)
 }
 
 func (r RowRange) proto() *btpb.RowSet {
-	rr := &btpb.RowRange{
-		StartKey: &btpb.RowRange_StartKeyClosed{StartKeyClosed: []byte(r.start)},
+	rr := &btpb.RowRange{}
+
+	switch r.startBound {
+	case rangeOpen:
+		rr.StartKey = &btpb.RowRange_StartKeyOpen{StartKeyOpen: []byte(r.start)}
+		break
+	case rangeClosed:
+		rr.StartKey = &btpb.RowRange_StartKeyClosed{StartKeyClosed: []byte(r.start)}
+		break
+	case rangeUnbounded:
+		// leave unbounded
+		break
 	}
-	if !r.Unbounded() {
-		rr.EndKey = &btpb.RowRange_EndKeyOpen{EndKeyOpen: []byte(r.limit)}
+
+	switch r.endBound {
+	case rangeOpen:
+		rr.EndKey = &btpb.RowRange_EndKeyOpen{EndKeyOpen: []byte(r.end)}
+		break
+	case rangeClosed:
+		rr.EndKey = &btpb.RowRange_EndKeyClosed{EndKeyClosed: []byte(r.end)}
+		break
+	case rangeUnbounded:
+		// leave unbounded
+		break
 	}
+
 	return &btpb.RowSet{RowRanges: []*btpb.RowRange{rr}}
 }
 
 func (r RowRange) retainRowsAfter(lastRowKey string) RowSet {
 	if lastRowKey == "" || lastRowKey < r.start {
 		return r
 	}
-	// Set the beginning of the range to the row after the last scanned.
-	start := lastRowKey + "\x00"
-	if r.Unbounded() {
-		return InfiniteRange(start)
+
+	return RowRange{
+		// Set the beginning of the range to the row after the last scanned.
+		startBound: rangeOpen,
+		start:      lastRowKey,
+		endBound:   r.endBound,
+		end:        r.end,
+	}
+}
+
+func (r RowRange) retainRowsBefore(lastRowKey string) RowSet {
+	if lastRowKey == "" || (r.endBound != rangeUnbounded && r.end <= lastRowKey) {
+		return r
+	}
+
+	return RowRange{
+		startBound: r.startBound,
+		start:      r.start,
+		endBound:   rangeOpen,
+		end:        lastRowKey,
 	}
-	return NewRange(start, r.limit)
 }
 
 func (r RowRange) valid() bool {
-	return r.Unbounded() || r.start < r.limit
+	if r.Unbounded() {
+		return true
+	}
+
+	if r.startBound == rangeOpen || r.endBound == rangeOpen {
+		return r.start < r.end
+	} else if r.startBound == rangeClosed {
+		return r.start <= r.end
+	} else {
+		return true
+	}
```

Comment:
This pretty hard to reason about and is not immediately clear why this is correct. I would either heavily comment this or make it more explicit. ie

```
// if either end is unbounded, then the range is always valid
if r.Unbounded() {
   return true
}

// If either end is an open interval, then the start must be strictly less than the end
// And since neither end is unbounded, we don't have to check for empty strings
if r.startBound == rangeOpen || r.endBound == rangeOpen {
  return r.start < r.end
}

// At this point both endpoints must be closed, which makes [a,a] a valid interval
return r.start <= r.end
```

Reply:
Agreed. Thanks for the suggestion!